생성형 ai 도구 링크 : https://claude.ai/share/197d4528-d4ba-490d-b842-bfb07ff0489f

Q1

In [7]:
import re

text = "강의는 2026-05-06, 숙제는 2026-05-18에 마감"

(a)

In [8]:
re.match(r"\d+", text)

결과 : None / re.match()는 문자열의 처음 위치에서 패턴이 시작하는지 검사하는데, text는 처음이 \d가 아니므로 매칭에 실패한다.

(b)

In [9]:
re.search(r"\d{4}-\d{2}-\d{2}", text).group()

'2026-05-06'

결과 : '2026-05-06' / re.search는 문자열 어디든 처음 매칭되는 위치를 찾고, .group()으로 매칭된 문자열을 얻는다. 

(c)

In [11]:
re.findall(r"\d{4}-\d{2}-\d{2}", text)

['2026-05-06', '2026-05-18']

결과 : ['2026-05-06', '2026-05-18'] / 캡처 그룹이 없으므로 매칭된 전체 문자열들의 리스트를 반환한다. 

(d)

In [12]:
re.findall(r"(\d{4})-(\d{2})-(\d{2})", text)

[('2026', '05', '06'), ('2026', '05', '18')]

결과 : [('2026', '05', '06'), ('2026', '05', '18')] / 캡처 그룹이 둘 이상이므로 튜플 리스트를 반환한다.

(e)

In [13]:
re.findall(r"(?:\d{4})-(?:\d{2})-(?:\d{2})", text)

['2026-05-06', '2026-05-18']

결과 : ['2026-05-06', '2026-05-18'] / 비캡처 그룹으로, 묶기만 하고 기억하지 않으므로 캡처 그룹이 없는 (c)와 같은 결과를 반환한다.

추가 질문 :  캡처 그룹(...)은 매칭된 부분 문자열을 기억하고, 비캡처 그룹(?:...)은 묶기만 하고 기억은 하지 않는다. re.findall()은 캡처 그룹이 있으면 그룹만 돌려주고, 그룹이 둘 이상이면 튜플 리스트를 돌려주므로 (d)에선 캡처 그룹이 3개라 튜플 리스트가 반환된다. (c)는 그룹이 없고, (e)는 비캡처 그룹이므로 둘다 캡처 그룹이 없어 매칭된 전체 문자열을 원소로 하는 리스트를 반환한다. 

Q2

In [14]:
import re

html = "<b>안녕</b> <i>세상</i>!"
nums = "수강생 30명, 조교 3명"

(a)

In [15]:
re.sub(r"<.+>", "[T]", html)

'[T]!'

결과 : '[T]!' / .+ 는 탐욕적 수량자이므로 첫 <부터 마지막>까지 한 덩어리로 매칭한다. 

(b)

In [16]:
re.sub(r"<.+?>", "[T]", html)

'[T]안녕[T] [T]세상[T]!'

결과 : '[T]안녕[T] [T]세상[T]!' / .+? 는 게으른 수량자이므로 >를 만나는 즉시 매칭이 종료된다. 

(c)

In [17]:
re.sub(r"<[^>]+>", "[T]", html)

'[T]안녕[T] [T]세상[T]!'

결과 : '[T]안녕[T] [T]세상[T]!' / [^>]+ 는 >가 아닌 문자들만 매칭하는 부정 문자 클래스이다.

(d)

In [18]:
re.sub(r"(\d+)", r"<\1>", nums)

'수강생 <30>명, 조교 <3>명'

결과 : '수강생 <30>명, 조교 <3>명' / 원시 문자열 r"<\1>"을 썼으므로 정규표현식 엔진이 \1을 첫 번째 그룹으로 보고 매칭된 숫자로 치환하는 역참조가 동작한다.

(e)

In [19]:
re.sub(r"(\d+)", "<\1>", nums)

'수강생 <\x01>명, 조교 <\x01>명'

결과 : '수강생 <\x01>명, 조교 <\x01>명' / \1이 8진수 이스케이프로 해석되므로 역참조가 동작하지 않고 \x01로 치환된다.

추가 질문 (i) : (a)는 '.+'를 사용한 탐욕적 수량자로, 가능한 한 많이 매칭해서 첫 < 부터 마지막 > 까지 전부를 삼킨 후 이를 한 덩어리로 치환하지만, (b)는 '.+?'를 사용한 게으른 수량자로, 가능한 한 적게 매칭해서 >를 만나는 즉시 매칭을 끝내 네 개의 태그가 각각 따로 매칭된 후 치환된다. 

추가 질문 (ii) : (d)는 원시 문자열 r"<\1>"을 사용했으므로 정규표현식 엔진이 \1을 첫 번째 그룹으로 보고 매칭된 숫자 30과 3으로 치환하지만(역참조), (e)는 r을 뺀 "<\1>"을 사용하므로 파이썬이 \1을 8진수 이스케이프(\x01)로 해석하고, 그 결과 숫자 자리에 \x01이 들어간 값이 반환된다(역참조 동작 안함).

Q3

re.compile 로 미리 컴파일하기

In [7]:
import re
from collections import Counter

URL_RE = re.compile(r"https?://\S+")
TAG_RE = re.compile(r"<[^>]+>")
EMAIL_RE = re.compile(r"[\w.+-]+@[\w.-]+\.\w+")
PHONE_RE = re.compile(r"\d{2,4}-\d{3,4}-\d{4}")
MENTION_OR_TAG_RE = re.compile(r"[@#]\w+")
JAMO_RE = re.compile(r"[\u3131-\u3163]+")
WS_RE = re.compile(r"\s+")
HASHTAG_RE = re.compile(r"#(\w+)")

(a)

In [3]:
def clean_post(post: str) -> str:
    post = URL_RE.sub(" ", post)
    post = TAG_RE.sub("", post)
    post, _ = EMAIL_RE.subn("[이메일]", post)
    post, _ = PHONE_RE.subn("[전화]", post)
    post = MENTION_OR_TAG_RE.sub(" ", post)
    post = JAMO_RE.sub("", post)
    post = WS_RE.sub(" ", post).strip()
    return post

(b)

In [4]:
def extract_hashtags(post: str) -> list[str]:
    return HASHTAG_RE.findall(post)

(c)

In [5]:
def analyze_posts(posts: list[str]) -> dict:
    
    posts_n: int = len(posts)

    cleaned = [clean_post(post) for post in posts]
    avg_length: float = round(sum(len(c) for c in cleaned) / posts_n, 2)

    all_tags: list[str] = []
    for post in posts:
        all_tags.extend(extract_hashtags(post))
    hashtag_counts: dict[str, int] = dict(Counter(all_tags).most_common())

    masked_count: int = 0
    for post in posts:
        _, n_email = EMAIL_RE.subn("[이메일]", post)
        _, n_phone = PHONE_RE.subn("[전화]", post)
        masked_count += n_email + n_phone

    return {
        "posts_n": posts_n,
        "avg_length_after_clean": avg_length,
        "hashtag_counts": hashtag_counts,
        "masked_count": masked_count,
    }


In [11]:
posts: list[str] = [
    "오늘 #파이썬 수업 진짜 재밌었음!! @prof_kim @hong 감사 ㅎㅎ"
    "자료: https://etl.snu.ac.kr/lec17",
    "@lee @park 팀플 어디서 모이지ㅠㅠ #DCCP2026 #팀플 카톡 ㄱㄱ",
    "<b>중요</b>: 다음 시험 범위는 1-15강. "
    "문의는 mam3b@snu.ac.kr (010-1234-5678)로!",
    "   여러     공백과\n\n\n줄바꿈이     많은 텍스트   ",
    "ㅋㅋㅋ #파이썬 진짜 좋다 #추천 https://snu.ac.kr",
]

for post in posts:
    print(clean_post(post))

analyze_posts(posts)

오늘 수업 진짜 재밌었음!! 감사 자료:
팀플 어디서 모이지 카톡
중요: 다음 시험 범위는 1-15강. 문의는 [이메일] ([전화])로!
여러 공백과 줄바꿈이 많은 텍스트
진짜 좋다


{'posts_n': 5,
 'avg_length_after_clean': 19.4,
 'hashtag_counts': {'파이썬': 2, 'DCCP2026': 1, '팀플': 1, '추천': 1},
 'masked_count': 2}

(1) 각 게시물에 대한 clean_post의 반환값 5개 : 1. 오늘 수업 진짜 재밌었음!! 감사 자료:
2. 팀플 어디서 모이지 카톡
3. 중요: 다음 시험 범위는 1-15강. 문의는 [이메일] ([전화])로!
4. 여러 공백과 줄바꿈이 많은 텍스트
5. 진짜 좋다
(2) analyze_posts(posts)의 반환 딕셔너리 : {'posts_n': 5,
 'avg_length_after_clean': 19.4,
 'hashtag_counts': {'파이썬': 2, 'DCCP2026': 1, '팀플': 1, '추천': 1},
 'masked_count': 2}

6단계의 순서를 지켜야 하는 까닭 : \w는 모든 단어 문자를 포함하므로, 3단계와 4단계의 순서를 바꾸면 mam3b@snu.ac.kr 안의 @snu 가 4단계의 @\w+ 에 먼저 잡혀 공백으로 치환되고, 이로 인해 이메일이 산산조각나서 3단계의 이메일 마스킹이 실패한다. 또한 만약 4단계가 1단계보다 먼저 온다면 URL 내부의 # 이나 @가 포함된 부분이 해시태그로 오인되어 잘려나갈 수 있으며, 공백은 각 단계에서 계속 만들어질 수 있으므로 공백 정리를 수행하는 6단계는 마지막에 오는 것이 효율적이다. 